python/ray/_private/runtime_env/py_modules.py

```diff
@@ -80,16 +84,25 @@ def upload_py_modules_if_needed(
             module_uri = module_path
         else:
             # module_path is a local path.
-            excludes = runtime_env.get("excludes", None)
-            module_uri = get_uri_for_directory(module_path, excludes=excludes)
-            upload_package_if_needed(
-                module_uri,
-                scratch_dir,
-                module_path,
-                excludes=excludes,
-                include_parent_dir=True,
-                logger=logger,
-            )
+            if Path(module_path).is_dir():
+                excludes = runtime_env.get("excludes", None)
+                module_uri = get_uri_for_directory(module_path, excludes=excludes)
+                upload_package_if_needed(
+                    module_uri,
+                    scratch_dir,
+                    module_path,
+                    excludes=excludes,
+                    include_parent_dir=True,
+                    logger=logger,
+                )
+            elif Path(module_path).suffix == ".whl":
+                module_uri = get_uri_for_package(Path(module_path))
+                upload_package_to_gcs(module_uri, Path(module_path).read_bytes())
```

Comment:
seems like these two codepaths could be further unified. for example, `upload_package_if_needed` checks if the URI already exists (which we probably want to do here too?) and calls into `upload_package_to_gcs`. Something more unified might look like:

```
uri, local_path = None, None
if zip:
    uri = get_uri_for_dir(...)
    local_path = zip_local_dir(...)
elif whl:
    uri = get_uri_for_package(...)
    local_path = module_path

if not package_exists(uri):
    upload_package_to_gcs(uri, local_path)

Reply:
Sounds good, let me take a closer look at the "if_needed" part, I forgot what I had concluded there but it does seem like we should be doing it for `whl` also.